torch/csrc/jit/python/pybind_utils.h

```diff
@@ -612,7 +633,7 @@ inline IValue toIValue(
           py::cast<c10::intrusive_ptr<CustomClassHolder>>(obj));
     }
     case TypeKind::FutureType: {
-      return obj.cast<PythonFutureWrapper>().fut;
+      return obj.cast<std::shared_ptr<PythonFutureWrapper>>()->fut;
```

Comment:
I assume this cast has no impact on the PY refcnt, instead, the refcnt is handled by the shared_ptr?

Reply:
First the object is a C++ heap object whose lifetime and destructing time is managed by shared_ptr. Out of all shared_ptr instance, there is one that is constructed by pybind and passed to Python. Python has its own way of doing ref counting, when the python ref count is 0, it delete the one shared_ptr. 

The casting has impact on C++ ref counting, if its a unique_ptr!
Given a unique_ptr, it has to copy the data structure, for the cast() function, which takes a copy rather than a reference, to use. 
As I have discussed with @wanchaol in https://github.com/pytorch/pytorch/pull/35154#discussion_r396693593.

I want to make sure this is always the ref cnt pattern for data structures used byRPC system, where we don't want to have extra copies.